### ISSUE TITLE
Fixing controller_manager

### ISSUE THREAD

**Issue by vic-bart** [MonashNovaRover/nova#1]
```
[robot_state_publisher-1] [INFO] [1726724341.457854064] [robot_state_publisher]: Robot initialized
[ros2_control_node-4] [INFO] [1726724342.028612494] [controller_manager]: Subscribing to '/robot_description' topic for robot description.
[ros2_control_node-4] [INFO] [1726724342.028781077] [controller_manager]: update rate is 20 Hz
[ros2_control_node-4] [WARN] [1726724342.031856311] [controller_manager]: Could not enable FIFO RT scheduling policy. Consider setting up your user to do FIFO RT scheduling. See [https://control.ros.org/master/doc/ros2_control/controller_manager/doc/userdoc.html] for details.
[ros2_control_node-4] [INFO] [1726724342.053654952] [controller_manager]: Received robot description from topic.
[ros2_control_node-4] [INFO] [1726724342.079027358] [controller_manager.resource_manager]: Loading hardware 'GazeboSystem' 
[ros2_control_node-4] [ERROR] [1726724342.079163071] [controller_manager.resource_manager]: Exception while loading hardware: According to the loaded plugin descriptions the class gazebo_ros2_control/GazeboSystem with base class type hardware_interface::SystemInterface does not exist. Declared types are  blcmd_hardware/BLCMDHardware mock_components/GenericSystem
[ros2_control_node-4] [WARN] [1726724342.079258798] [controller_manager]: Could not load and initialize hardware. Please check previous output for more details. After you have corrected your URDF, try to publish robot description again.
```
controller_manager is failing on startup due to not being able to load the gazebo_ros2_control/GazeboSystem plugin

**Comment by vic-bart** [comment]
![image](https://github.com/user-attachments/assets/418666c7-df74-45b2-8e7f-811daadc9af9)
I think this means that the code can't find the source file for the gazebo plugins (since the path is empty/undeclared), but what's the correct filepath? This [answer](https://answers.gazebosim.org/question/13391/how-to-set-gazebo_plugin_path-correctly-and-add-the-plugin-into-gazebo_ros/) seems to explain what needs to be done, but where is our source file? 
![image](https://github.com/user-attachments/assets/142b693d-4344-4c96-ac8c-5bbef82d2ec5)
Looking at the nix store, I think it might be one of these?

**Comment by vic-bart** [comment]
No it's not that, because this issue isn't present on the Metabox's master build, which also has an empty `$GAZEBO_PLUGIN_PATH`

**Comment by vic-bart** [comment]
[Wait a minute....](https://control.ros.org/jazzy/doc/gazebo_ros2_control/doc/index.html)
![image](https://github.com/user-attachments/assets/a8bb298e-f961-422b-8afe-a35e65f2452f)
It can't be this surely; we tested auto_bringup with ros jazzy and everything worked fine, no issues

**Comment by vic-bart** [comment]
Context on the issue:
The controllers joint_broad and pivot_drive are failing because they're waiting on controller_manager
![image](https://github.com/user-attachments/assets/29dd0ad7-fb94-47eb-a976-aa6bb5c4ce8d)
Running the controller_manager node separately (it's the ros2_control_node executable from controller_manager ROS package) gives the specific error message (which is likely why the controller_manager is silently failing).
![image](https://github.com/user-attachments/assets/d20d2e11-bde2-41fc-9b44-1f47faf97f39)

**Comment by vic-bart** [comment]
Tried tracking down the error, and found that controller_manager from ros2_control calls another node, resource_manager, which has the following [code](https://github.com/ros-controls/ros2_control/blob/84e85f9f0bbd67af9eebbe40bf5eef89792945ff/hardware_interface/src/resource_manager.cpp#L37):
```cpp
template <class HardwareT, class HardwareInterfaceT>
  [[nodiscard]] bool load_hardware(
    const HardwareInfo & hardware_info, pluginlib::ClassLoader<HardwareInterfaceT> & loader,
    std::vector<HardwareT> & container)
  {
    bool is_loaded = false;
    try
    {
      RCLCPP_INFO(get_logger(), "Loading hardware '%s' ", hardware_info.name.c_str());
      // hardware_plugin_name has to match class name in plugin xml description
      auto interface = std::unique_ptr<HardwareInterfaceT>(
        loader.createUnmanagedInstance(hardware_info.hardware_plugin_name));
```
This matches the terminal logs, and suggests that the error originates from the createUnmanagedInstance function of pluginlib::ClassLoader. Looking at its [code](https://github.com/ros/pluginlib/blob/jazzy/pluginlib/include/pluginlib/class_loader.hpp), there is only the following:
```hpp
T * createUnmanagedInstance(const std::string & lookup_name);

  /// Return a list of all available plugin manifest paths for this ClassLoader's base class type.
  /**
   * \return A vector of strings corresponding to the paths of all available plugin manifests
   */
```
But, at the bottom of the file, it inherits the implementations from a different file!
```hpp
// Note: The implementation of the methods is in a separate file for clarity.
#include "./class_loader_imp.hpp"
```
And looking in this [file](https://github.com/ros/pluginlib/blob/jazzy/pluginlib/include/pluginlib/class_loader_imp.hpp), we find the method:
```hpp
template<class T>
T * ClassLoader<T>::createUnmanagedInstance(const std::string & lookup_name)
/***************************************************************************/
{
  RCUTILS_LOG_DEBUG_NAMED("pluginlib.ClassLoader",
    "Attempting to create UNMANAGED instance for class %s.",
    lookup_name.c_str());

  if (!isClassLoaded(lookup_name)) {
    loadLibraryForClass(lookup_name);
  }

  T * instance = 0;
  try {
    RCUTILS_LOG_DEBUG_NAMED("pluginlib.ClassLoader",
      "Attempting to create instance through low level multi-library class loader.");
    std::string class_type = getClassType(lookup_name);
    RCUTILS_LOG_DEBUG_NAMED("pluginlib.ClassLoader", "%s maps to real class type %s",
      lookup_name.c_str(), class_type.c_str());
    instance = lowlevel_class_loader_.createUnmanagedInstance<T>(class_type);
    RCUTILS_LOG_DEBUG_NAMED("pluginlib.ClassLoader",
      "Instance of type %s created.",
      class_type.c_str());
  } catch (const class_loader::CreateClassException & ex) {
    RCUTILS_LOG_DEBUG_NAMED("pluginlib.ClassLoader",
      "Exception raised by low-level multi-library class loader when attempting "
      "to create UNMANAGED instance of class %s.",
      lookup_name.c_str());
    throw pluginlib::CreateClassException(ex.what());
  }
  return instance;
}
```
Which calls a method belonging to lowlevel_class_loader_, which is actually defined back in the other [file](https://github.com/ros/pluginlib/blob/jazzy/pluginlib/include/pluginlib/class_loader.hpp):
```hpp
private:
  std::vector<std::string> plugin_xml_paths_;
  // Map from library to class's descriptions described in XML.
  std::map<std::string, ClassDesc> classes_available_;
  std::string package_;
  std::string base_class_;
  std::string attrib_name_;
  class_loader::MultiLibraryClassLoader lowlevel_class_loader_;  // The underlying classloader
};
```
And class_loader is included at the top of this file:
```
#include "class_loader/multi_library_class_loader.hpp"
```
So tracking that down, we find this [code](https://github.com/ros/class_loader/blob/jazzy/include/class_loader/multi_library_class_loader.hpp):
```hpp
  /**
   * @brief Creates an instance of an object of given class name with ancestor class Base
   * This version does not look in a specific library for the factory, but rather the first open
   * library that defines the class
   * This version should not be used as the plugin system cannot do automated safe loading/unloadings
   *
   * @param Base - polymorphic type indicating base class
   * @param class_name - the name of the concrete plugin class we want to instantiate
   * @return An unmanaged Base* to newly created plugin
   */
  template<class Base>
  Base * createUnmanagedInstance(const std::string & class_name)
  {
    ClassLoader * loader = getClassLoaderForClass<Base>(class_name);
    if (nullptr == loader) {
      throw class_loader::CreateClassException(
              "MultiLibraryClassLoader: Could not create class of type " + class_name);
    }
    return loader->createUnmanagedInstance<Base>(class_name);
  }

  /**
   * @brief Creates an instance of an object of given class name with ancestor class Base
   * This version takes a specific library to make explicit the factory being used
   * This version should not be used as the plugin system cannot do automated safe loading/unloadings
   *
   * @param Base - polymorphic type indicating Base class
   * @param class_name - name of class for which we want to create instance
   * @param library_path - the fully qualified path to the runtime library
   */
  template<class Base>
  Base * createUnmanagedInstance(const std::string & class_name, const std::string & library_path)
  {
    ClassLoader * loader = getClassLoaderForLibrary(library_path);
    if (nullptr == loader) {
      throw class_loader::NoClassLoaderExistsException(
              "Could not create instance as there is no ClassLoader in MultiLibraryClassLoader "
              "bound to library " + library_path +
              " Ensure you called MultiLibraryClassLoader::loadLibrary()");
    }
    return loader->createUnmanagedInstance<Base>(class_name);
  }
```
At this point, I think I got a bit lost on a tangent, so I went back to the error, which explicitly states that the supposed plugin it's meant to instantiate doesn't exist &mdash; a check that would most likely happen earlier on, leading me back to this [file](https://github.com/ros/pluginlib/blob/jazzy/pluginlib/include/pluginlib/class_loader_imp.hpp):
```hpp
  if (!isClassLoaded(lookup_name)) {
    loadLibraryForClass(lookup_name);
  }
```
Which is defined further down in the file:
```hpp
template<class T>
bool ClassLoader<T>::isClassLoaded(const std::string & lookup_name)
/***************************************************************************/
{
  return lowlevel_class_loader_.isClassAvailable<T>(getClassType(lookup_name));
}
```
So going back to the multi_library_class_loader, we find this [code](https://github.com/ros/class_loader/blob/jazzy/include/class_loader/multi_library_class_loader.hpp):
```hpp
  /**
   * @brief Indicates if a class has been loaded and can be instantiated
   *
   * @param Base - polymorphic type indicating Base class
   * @param class_name - name of class that is be inquired about
   * @return true if loaded, false otherwise
   */
  template<class Base>
  bool isClassAvailable(const std::string & class_name) const
  {
    std::vector<std::string> available_classes = getAvailableClasses<Base>();
    return available_classes.end() != std::find(
      available_classes.begin(), available_classes.end(), class_name);
  }
```
getAvailableClasses is defined in the same file, further down:
```hpp
  /**
   * @brief Gets a list of all classes that are loaded by the class loader
   *
   * @param Base - polymorphic type indicating Base class
   * @return A vector<string> of the available classes
   */
  template<class Base>
  std::vector<std::string> getAvailableClasses() const
  {
    std::vector<std::string> available_classes;
    for (auto & loader : getAllAvailableClassLoaders()) {
      std::vector<std::string> loader_classes = loader->getAvailableClasses<Base>();
      available_classes.insert(
        available_classes.end(), loader_classes.begin(), loader_classes.end());
    }
    return available_classes;
  }
```
getAllAvailableClassLoaders is defined at the bottom of this file, but not implemented:
```hpp
  /**
   * @brief Gets all class loaders loaded within scope
   *
   * @return vector with available ClassLoader pointers
   */
  ClassLoaderVector getAllAvailableClassLoaders() const;
```
The only implementation I could find was in this [file](https://github.com/ros/class_loader/blob/jazzy/src/multi_library_class_loader.cpp#L109):
```cpp
ClassLoaderVector MultiLibraryClassLoader::getAllAvailableClassLoaders() const
{
  ClassLoaderVector loaders;
  for (auto & it : impl_->active_class_loaders_) {
    loaders.push_back(it.second);
  }
  return loaders;
}
```
I'm not 100% sure how hpp and cpp files interact &mdash; this cpp file is never "included" or otherwise referred to, so does this implementation affect the hpp file? I don't know. Either way, the function of getAllAvailableClassLoaders is pretty self-explanatory.

What seems to be happening is that controller_manager spins up resource_manager, which during initialisation, attempts to load the hardware interfaces specified in robot_description. However, when checking if the class is loaded/available, it iterates through all available classes/class_loaders, and doesn't found our plugin (the class gazebo_ros2_control/GazeboSystem with base class type hardware_interface::SystemInterface) loaded.

So I think I can confidently say that the issue is not with ros2_control, but rather with whatever is supposed to be loading our plugin (which no longer is doing so).

**Comment by vic-bart** [comment]
Trying to narrow down exactly when/how this error occurred, so I checked out nixfiles at commit a155b68 and rover at commit b1a4a11f (before upgrading to ros jazzy), and built by specifying `-A pkgs.rosPackages.nova-workspace.jazzy`, and this error was STILL present, which leads me to believe that the problem is with gazebo_ros2_control.

So I checked out [gazebo_ros2_control](https://github.com/ros-controls/gazebo_ros2_control), and say the released a new version (0.7.3) very recently, so I changed the overlay in nixfiles to build that instead (added the correct hash), and checked the build result in the nix-store to confirm it was the newer version. The same error was present. 

ros2_control and gazebo_ros2_control is up-to-date, and this error is present. Usually I'd make an issue, but gazebo_ros2_control isn't even supposed to run with jazzy, so we probably won't get helped. 

Since we're already planning to move from Gazebo Classic to Gazebo Harmonic, and I think it's a better use of our time if we start that transition immediately rather than waste time trying to resolve this issue, and then transition later on anyway. Will start preliminary work to gauge the difficulty of the proposed transition, which will be continued in [this issue](https://github.com/MonashNovaRover/nova/issues/3).
